### PR TITLE
Make the translation function fully type-safe

### LINKED ISSUE #1811: TFunction casts no longer work - remove 4.1 typescript workaround?
## 🐛 Bug Report

A workaround was introduced in https://github.com/i18next/react-i18next/pull/1227 by [pedrodurek](https://github.com/pedrodurek)
due to bug https://github.com/i18next/react-i18next/issues/1222 reported by @jigsawye.

The underlying issue https://github.com/microsoft/TypeScript/issues/41406 has been fixed.

I noticed because we were doing some very latent updates and when using react-i18next (not i18next directly), casting no longer typechecks as is tested in https://github.com/i18next/i18next/blob/master/test/typescript/t.test.ts#L16-L22

## To Reproduce

```js
function returnCasts(t: TFunction) {
  const s: string = t('friend'); // same as <string>
  const s2: string = t`friend`;
  const o: object = t<object>('friend');
  const sa: string[] = t<string[]>('friend');
  const oa: object[] = t<object[]>('friend');
}
```

## Expected behavior

The same code tested in i18next should work in react-i18next.

I believe this can be satisfied by removing the additional types for 4.1, but I haven't been keeping up with all the changes, so I could be wrong.

## Your Environment

- *runtime version*: browser
- *react-i18next version*: 11.18.3
- *i18next version*: 21.8.16
- *os*: Mac
- *typescript*: 4.7.4

### PR DESCRIPTION
**Pull request by pedrodurek**
Closes https://github.com/i18next/i18next/issues/1504
Closes #1811

We'll no longer support typescript below v4.1

#### Checklist

- [x] only relevant code is changed (make a diff before you submit the PR)
- [ ] run tests `npm run test`
- [x] tests are included

#### Checklist (for documentation change)

- [ ] only relevant documentation part is changed (make a diff before you submit the PR)
- [ ] motivation/reason is provided

### CODE REVIEW

**Comment by pedrodurek** [comment]
It's still pending to:

- [ ]  Add some tests for the new structure
- [ ] Add some examples
- [ ] Update documentation
- [ ] Specify the breaking changes

**Comment by pedrodurek** [comment]
I'll update the remaining work this week

**Comment by JohnHallstrom** [comment]
Is there an ETA on this?

**Comment by JohnHallstrom** [comment]
> Hey @JohnHallstrom, I'm in the middle of the a moving houses process, but I intend to have everything in place by end of next week. Sorry for the delay ❤️

That sounds great! I hope everything is going well with your move until then. Cheers!

**Comment by luisbritosuper** [comment]
This would be great! Any news or ETA?

**Comment by obritoluis** [comment]
This would be great! Any news or ETA?

**Comment by CanRau** [comment]
Please dear stale bot, could you kindly reconsider your decision 🥺

**Comment by adrai** [comment]
> Please dear stale bot, could you kindly reconsider your decision 🥺

@pedrodurek you're on it, right?

**Comment by pedrodurek** [comment]
Hey @adrai, yes I am, it's been hard to find some time answer messages and address some issues lately, but the PR is good now, so we'll be able to merge it this week or the next one.
I'm just finishing the typescript doc right now.

**Comment by CanRau** [comment]
It's still in process Staley, a little more patience please 🙏

**Comment by CanRau** [comment]
🥺

**Comment by pedrodurek** [comment]
Hey @adrai, it should be a `Minor` release. It'll probably not have any breaking changes, but I expect some devs to report bugs and/or limitations.

File: test/typescript/t.test.ts [hunk 1]
```diff
@@ -16,9 +16,9 @@ function overloadedUsage(t: TFunction) {
 function returnCasts(t: TFunction) {
   const s: string = t('friend'); // same as <string>
   const s2: string = t`friend`;
-  const o: object = t<object>('friend');
-  const sa: string[] = t<string[]>('friend');
-  const oa: object[] = t<object[]>('friend');
+  const o: object = t<string, object>('friend');
+  const sa: string[] = t<string, string[]>('friend');
+  const oa: object[] = t<string, object[]>('friend');
```
Comment (rosskevin) [inline test/typescript/t.test.ts:21]: I don't disapprove of this, but it should be noted that this technically is a type-breaking change.

File: tsconfig.json [hunk 2]
```diff
@@ -15,5 +15,5 @@
     "allowSyntheticDefaultImports": true
   },
   "include": ["./index.d.ts", "./test/**/*.ts*"],
-  "exclude": ["test/typescript/nonEsModuleInterop/**/*.ts"]
+  "exclude": ["test/typescript/nonEsModuleInterop/**/*.ts", "test/typescript/custom-types"]
```
Comment (rosskevin) [inline tsconfig.json:18]: I see that the `custom-types` `tsconfig.json` was added with a `t.test.ts`, but I don't see where it is being tested in one of the scripts e.g.:

```json
    "test:typescript": "tslint --project tsconfig.json",
    "test:typescript:noninterop": "tslint --project tsconfig.nonEsModuleInterop.json",
```

Am I missing something?

**Review by rosskevin — CHANGES_REQUESTED** [review]

File: test/typescript/t.test.ts (hunk 1 above)
Comment (pedrodurek) [reply test/typescript/t.test.ts:21]: yeah, I agree. Just want to reinforce that it should be a `Minor` release, not only a `Patch` release

File: tsconfig.json (hunk 2 above)
Comment (pedrodurek) [reply tsconfig.json:18]: Yeah, I missed that, thanks!
The IDE will point out the error, but it won't fail on the CI, I'll update the package json

File: test/typescript/t.test.ts (hunk 1 above)
Comment (rosskevin) [reply test/typescript/t.test.ts:21]: By semver requirements, breaking would be `major`.  I don't have a strong preference one way or the other, I just worry about those that do expect strict adherence to semver.

**Review by rosskevin — APPROVED** [review]

**Comment by adrai** [comment]
I would go with a major bump...
btw: is there something to be done in other repos, like react-i18next?

**Comment by rosskevin** [comment]
Yes, #1811 is actually about `react-i18next`, with that code move to this PR.  The re-definition of `TFunction` should be removed there.

**Comment by adrai** [comment]
So if ok for you, I will merge this PR tomorrow morning and also create a new release.
Would be happy if you could check also for react-i18next and next-i18next afterwards. 🙏

**Comment by rosskevin** [comment]
@pedrodurek is already on it for react-i18next https://github.com/i18next/react-i18next/pull/1501

I checked next-i18next and it does not need any changes.

**Comment by adrai** [comment]
included in v22.0.0

**Pull request merged**